### ISSUE TITLE
Pattern / Wildcard not expanded

### ISSUE THREAD

**Issue by LhKipp** [nushell/nushell#2514]
**Describe the bug**
A wildcard / pattern is not expanded.

**To Reproduce**
1. cd /
2. echo * #Pattern not expanded
3. rm * -rf #Pattern expanded

**Expected behavior**
Wildcards should always be expanded.

**Configuration (please complete the following information):**
 - OS (e.g. Windows): Arch Linux
 - Nu version (you can use the `version` command to find out): 0.19.0
 - Optional features (if any):

I think it has to do with the commands Signatures. Some declare a argument of SyntaxShape::Pattern, others don't.
From a user perspective, I always want wildcards expanded. Having it expand for some commands, for others not is highly confusing. From a script/command writer perspective I want to declare which kind of expressions my command takes and how many. I don't want to handle patterns.

**Comment by thegedge** [comment]
I haven't 100% settled on whether or not we should expand globs yet, but I'm leaning far more towards not expanding. I have a lot of thoughts here, so bear with me 😅 

## Command/script writers

On the one hand, I definitely can see that it's not as fun for the command/script writer to have to think about globs. Technically, command writers are thinking about it though, given they can specify `SyntaxShape::Pattern` in signatures. We can always make it really easy for command writers to expand globs. Since commands are written infrequently, but consumed frequently, I'm not as worried here. I take a roughly similar stance for script writers, but if we really wanted to simplify this, we could consider syntax for expanding a glob (see "Eager pattern expansion" below).

On the other hand, it fits better with nushell's streaming philosophy to have them do so.

What I mean by the "streaming approach" is this:

```
ls **/*.rs | first
```

If we have efficient streaming facilities for globs, we won't have to `readdir` (perhaps even `stat`, depending on the glob) more than we need. In the above example, we'll stop after we find the first match. Mind you, performance isn't a core value for nushell, but I thought I'd mention this. Another example:

```
if $true { echo test } { echo **/* } | first
```

At what point should we expand `**/*`? (see the expansion section below for some discussion on this).

## Users

> From a user perspective, I always want wildcards expanded. Having it expand for some commands, for others not is highly confusing.

To start, your example `echo *` could be expressed as `ls | get name` currently. Personally, I would much rather see `ls | get name` in a script than `echo *`. I find it far more readable. In this specific case, I disagree with you and would not want it expanded.

In terms of it being confusing, I think perhaps it's confusing when coming from other shells where such expansion exists. As you mentioned, we have `SyntaxShape::Pattern`, so a glob pattern is a typed piece of data that commands can consume. What if I had a command that took an arbitrary number of patterns and answered the question "does `$it` match any of these globs?"

```
<pipeline> | path matches *.txt *csv
```

If we expand the glob, that simply doesn't work (unless `$it` was in the current working directory). To force it to work we'd have to use `SyntaxShape::String`, but I see that as an unfortunate compromise since we're not being as expressive as the type system allows. We'd end up having the glob syntax check happen at runtime, which isn't terrible, but what if `<pipeline>` was very expensive? I'd much rather show errors sooner rather than later in that case. It really hits our core philosophy of nushell being a place for growing from experiments to programs.

## Other thoughts

I think the most interesting case of glob expansion is for externals, because externals are outside of the nushell ecosystem and are unlikely to be built to work with globs. For example:

```
wc *.txt
```

This doesn't work (unless you have a file literally named `*.txt`). We really want to do our best to work with externals though because without them nushell would not be nearly as useful. Possible options:

1. eager expansion (described below),
2. `SyntaxShape::Path` expansion (described below), or
3. expect people to write this in some nushell-like form. One option is `ls *.txt | wc $it.name`, which will spawn `wc` a bunch (slow). Another option is `wc $(ls **/*.txt | get name)`

(3) works today, so I'm inclined to defer making a decision here until there's a clear "we should definitely do this".

## `SyntaxShape::Path` expansion

Suppose we have `rest(SyntaxShape::Path)`, which indicates we want a single path, we could consider expanding a bare (bare = `*.txt", versus quoted `"*.txt"`, which would be a string) pattern.

Note, this expansion has to be based on the active `Shell`. We'd either have to expand before executing a command, or inject a shell into the parser (in other words, the parser would expose a trait for expanding globs).

This is worthy of a larger discussion, perhaps even an RFC. Should we expand `*.txt` for `String`? Should we expand `Range` for `Int` / `Number`?

## Eager glob expansion

We could consider syntax for expanding a `SyntaxShape::Pattern` into a list of names though. For example, `<*.txt>` could be equivalent to `$(ls *.txt | get name)`. This is just an example though, so we'd want to work on better syntax. In general, expanding the syntax is the thing I'm most concerned about, but since we're not 1.0 we can always try things out and drop it if we later find there are better alternatives.

**Comment by LhKipp** [comment]
Let me note that you are using a glob pattern here:
```shell 
ls **/*.rs | first
```
as a set of filenames, while you are using a glob pattern in your example from below
```shell
<pipeline> | path matches *.txt *csv
```
for wildcard matching (Does the string $it match *.txt or \*csv (\* is to be interpreted as a wildcard)).

Both use cases are something totaly different from my pov. However they share syntax.
IMO: Expanding it once as a set of filenames, and not expanding it once for string matching while both commands expect a SyntaxShape::Pattern is confusing, as SyntaxShape::Pattern is used in two different ways.

There is another use case for glob patterns:
```shell
$/root: cd **/nushell
```
In case there is ambiguity, cd would error out, as it only supports 1 argument.
Given that people may use glob patterns like that, I wouldn't know a command which should restrict its users with SyntaxShape::Path instead of using SyntaxShape::Pattern and asserting the len of the expanded pattern to be 1. 

In case one would automatically expand patterns, when the command has a signature, which expects a path (or multiple), the parser can do the length checking.


Given your good critics, I am now even doubting myself. One core concept is that any command can let the parser parse the input has he needs. Therefore strict rules like: "Always expand wildcards" might be the easiest to understand from a user perspective, but would limit the exploratory phase nushell is currently in.
However using SyntaxShape::Pattern in an ambiguous way, like in your example above is not the way to go IMO.
I would personaly (without putting to much thought into it) remove SyntaxShape::Pattern, and expand every pattern that is used in positions where a SyntaxShape::Path is expected. Additionaly there should be some syntax to expand a pattern. (Maybe `echo $(nushell/**)`?).



> This is worthy of a larger discussion, perhaps even an RFC. Should we expand *.txt for String? Should we expand Range for Int / Number?

True. You are free to close this PR. I do think an RFC may be a better plattform to discuss this.

**Comment by thegedge** [comment]
tl;dr I see two major points from our discussion.

- There doesn't seem to be a clearly compelling argument for one way or the other. They both have pros and cons 🙂
- Your proposal seems to be a combination of two things: drop `SyntaxShape::Pattern`, but still support globs by expanding them into a list of matching files, which would work for `SyntaxShape::Path`. If I got that right, the first question we should answer is "Should glob patterns be their own type?".

## Responding to your comment

> IMO: Expanding it once as a set of filenames, and not expanding it once for string matching while both commands expect a SyntaxShape::Pattern is confusing, as SyntaxShape::Pattern is used in two different ways.

I don't see it as confusing at all, when you come at this 

Let's consider the following rust functions:

```rust
fn ls(pattern: PatternType) -> Vec<Metadata>;
fn matches(path: Path, pattern: PatternType) -> bool;
```

It's up to me to understand how these functions work, and how they consume a `PatternType` argument. It's up to the compiler to ensure I give it a `PatternType`. That's the way I see the examples from my previous comment. If we were to continue with this Rust analogy, your recommendation of eliminating `Pattern` and expanding them into paths would perhaps look something like this:

```rust
fn ls(paths: impl Into<Vec<Path>>) -> Vec<Metadata>;
fn matches(path: Path, pattern: impl Into<String>) -> bool;
```

`ls` works fine in this case, but `matches` is now a) harder to understand, and b) prone to error, since I could perhaps provide an invalid string. We can check that at runtime, but I'm advocating for checking this statically to avoid unnecessary work.

> Additionaly there should be some syntax to expand a pattern. (Maybe echo $(nushell/**)?).

I'm not sure of the syntax, but I don't think it'll end up being `$(nushell/**)` as that basically says `nushell/**` is a valid pipeline. I gave an example above that I think would not be ambiguous (`<nushell/**>`), but not a huge fan of it (another idea is proposed below).

## Should globs be their own type?

So this is actually a great question, as it's one I've tossed around in my head lately. I really do like the idea of globs being a root type, because then (as mentioned earlier) we can statically verify the correctness of the globs.

For command writers there's an oddity: they have to choose between `SyntaxShape::Path` and `SyntaxShape::Pattern`, the latter being a superset of the former. This doesn't feel much different than a command writer using `String` though, since they could then use `glob::glob` or `Path::new` that string.

## Another idea!

I'm actually starting to think that what we have here is actually another example of a "splatting" operator. I'm thinking something like `expression...` performs an "expand" on the underlying expression. So, for example:

- `echo [1 2 3]...` → `echo 1 2 3`
- `echo **/*.txt...` → `echo a.txt subdir/b.txt`
- `echo 1..5...` → `echo 1 2 3 4 5`

Perhaps the behaviour could vary based on the position. For example, if `1..5...` was used for a flag that expects a table of integers, it would become `[1 2 3 4 5]`. If we keep `Pattern`, `*.txt...` would just stay `*.txt` if the flag was expecting a pattern. In the positional argument position it gets expanded based on the underlying shape.

Things to figure out in an RFC, if we can get consensus on this being a good idea 🙂

Ignoring the specific syntax of "splatting", what do you think of that idea?

**Comment by thegedge** [comment]
And just wanted to note, @LhKipp, I'm not trying to be all "NOOOO" here. I think there's a surprising amount of nuance here that I wanted to discuss, and I'm finding this discussion productive 🙂

**Comment by LhKipp** [comment]
> Should globs be their own type?
So this is actually a great question, as it's one I've tossed around in my head lately. I really do like the idea of globs being a root type, because then (as mentioned earlier) we can statically verify the correctness of the globs.

For a pattern provided as an argument instead of a path (or multiple paths), one can verify at evaluation time, that the pattern is expandable to at least n paths.
I am not sure what one could statically verify about a pattern used in a command like matches. Can you give some more details here?
How would you discriminate between a pattern used as an expansion for (multiple - including 0) path('s) or as a pattern to string match against, when statically checking for correctness?

> [edit]
Given the above, I'm arguing for also dropping Path if we were to drop Pattern, since Path::new is infallible (in other words, it's not really different from String). I actually would like to not have that, because I think types like these can be far more expressive. Maybe today we don't statically verify patterns, but we could easily add that (I've been playing around with path abstractions).

I agree. We should keep expressiveness.

SyntaxShape::Path has the advantage over SyntaxShape::String of beeing expressive. 
This is also true for SyntaxShape::Pattern vs SyntaxShape::String (IMO: if it is used as something to match against. In place of a path - I personally think - SyntaxShape::Pattern is more confusing in an shell environment than helpfull.)

The problem with SyntaxShape::Pattern vs SyntaxShape::Path is that Path is limiting. Therefore no one would use SyntaxShape::Path anymore I guess. Thats a real concern for me.

Maybe a possible solution might be: 
- Implementing a coercion rule for an Expression::Pattern to SyntaxShape::Path or Multiple SyntaxShape::Path. (Now SyntaxShape::Path will still be used)
- Keeping SyntaxShape::Pattern for commands like matches.

> There doesn't seem to be a clearly compelling argument for one way or the other. They both have pros and cons

Totally agree. Having this discussion enabled me to see the pros and cons of both solutions. It's fairly hard to say which way to go.

> Ignoring the specific syntax of "splatting", what do you think of that idea?

Having a uniform way of 'expanding' expressions sounds great! Not to sure about adding a table around it automatically in some situations. But the idea overall is worth looking into.

**Comment by thegedge** [comment]
Okay, perhaps we should explore the splatting idea first. Once we figure out splatting, we could have a configuration option to enable automatic splatting (e..g, expand a `Pattern` for any signature that has 1+ positional `Path` arguments).

**Comment by LhKipp** [comment]
> Okay, perhaps we should explore the splatting idea first. Once we figure out splatting, we could have a configuration option to enable automatic splatting (e..g, expand a Pattern for any signature that has 1+ positional Path arguments).

Agree!

> I'm mostly focused on the input being valid pattern syntax. If the pattern syntax were to ever expand to something more complex (for example, think about the expanded globs that zsh supports), this could become more important. For the basics – *, ?, and ** – it's less interesting. 

Thanks for pointing out the zsh extended globbing syntax :). Valid example!
